File: deepskilling/M2_Data Structures_and_Algorithms/InventoryManagementSystem.java
import java.util.HashMap;
import java.util.Map;

public class InventoryManagementSystem {
    static class Product {
        String productId;
        String productName;
        int quantity;
        double price;

        Product(String productId, String productName, int quantity, double price) {
            this.productId = productId;
            this.productName = productName;
            this.quantity = quantity;
            this.price = price;
        }

        @Override
        public String toString() {
            return String.format("Product{id=%s, name=%s, quantity=%d, price=%.2f}",
                    productId, productName, quantity, price);
        }
    }

    private final Map<String, Product> inventory = new HashMap<>();

    public void addProduct(Product product) {
        inventory.put(product.productId, product);
    }

    public boolean updateProduct(String productId, int quantity, double price) {
        Product product = inventory.get(productId);
        if (product == null) {
            return false;
        }
        product.quantity = quantity;
        product.price = price;
        return true;
    }

    public boolean deleteProduct(String productId) {
        return inventory.remove(productId) != null;
    }

    public Product getProduct(String productId) {
        return inventory.get(productId);
    }

    public void printInventory() {
        if (inventory.isEmpty()) {
            System.out.println("Inventory is empty.");
            return;
        }
        inventory.values().forEach(System.out::println);
    }

    public static void main(String[] args) {
        InventoryManagementSystem ims = new InventoryManagementSystem();

        ims.addProduct(new Product("P001", "Widget", 25, 9.99));
        ims.addProduct(new Product("P002", "Gadget", 40, 14.50));
        ims.addProduct(new Product("P003", "Sprocket", 15, 24.75));

        System.out.println("Initial inventory:");
        ims.printInventory();

        ims.updateProduct("P002", 35, 13.99);
        System.out.println("\nAfter updating P002:");
        System.out.println(ims.getProduct("P002"));

        ims.deleteProduct("P001");
        System.out.println("\nAfter deleting P001:");
        ims.printInventory();
    }
}

File: deepskilling/M2_Data Structures_and_Algorithms/InventoryManagementSystem.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class InventoryManagementSystem
{
    public class Product
    {
        public string ProductId { get; set; }
        public string ProductName { get; set; }
        public int Quantity { get; set; }
        public double Price { get; set; }

        public Product(string productId, string productName, int quantity, double price)
        {
            ProductId = productId;
            ProductName = productName;
            Quantity = quantity;
            Price = price;
        }

        public override string ToString()
        {
            return $"Product{{id={ProductId}, name={ProductName}, quantity={Quantity}, price={Price:F2}}}";
        }
    }

    private readonly Dictionary<string, Product> _inventory = new Dictionary<string, Product>();

    public void AddProduct(Product product)
    {
        _inventory[product.ProductId] = product;
    }

    public bool UpdateProduct(string productId, int quantity, double price)
    {
        if (!_inventory.TryGetValue(productId, out Product product))
        {
            return false;
        }
        product.Quantity = quantity;
        product.Price = price;
        return true;
    }

    public bool DeleteProduct(string productId)
    {
        return _inventory.Remove(productId);
    }

    public Product GetProduct(string productId)
    {
        _inventory.TryGetValue(productId, out Product product);
        return product;
    }

    public void PrintInventory()
    {
        if (_inventory.Count == 0)
        {
            Console.WriteLine("Inventory is empty.");
            return;
        }
        foreach (var product in _inventory.Values)
        {
            Console.WriteLine(product);
        }
    }

    public static void Main(string[] args)
    {
        InventoryManagementSystem ims = new InventoryManagementSystem();
        ims.AddProduct(new Product("P001", "Widget", 25, 9.99));
        ims.AddProduct(new Product("P002", "Gadget", 40, 14.50));
        ims.AddProduct(new Product("P003", "Sprocket", 15, 24.75));

        Console.WriteLine("Initial inventory:");
        ims.PrintInventory();

        ims.UpdateProduct("P002", 35, 13.99);

        Console.WriteLine("\nAfter updating P002:");
        Console.WriteLine(ims.GetProduct("P002"));

        ims.DeleteProduct("P001");

        Console.WriteLine("\nAfter deleting P001:");
        ims.PrintInventory();
    }
}
